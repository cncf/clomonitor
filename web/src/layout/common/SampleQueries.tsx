import { isUndefined, sampleSize } from 'lodash';
import { Fragment, memo } from 'react';
import { Link } from 'react-router-dom';

import prepareQueryString from '../../utils/prepareQueryString';

interface Props {
  className?: string;
  lineBreakIn?: number;
}

interface SampleQuery {
  name: string;
  filters: any;
}

const QUERIES: SampleQuery[] = [
  {
    name: 'Only graduated projects',
    filters: {
      pageNumber: 1,
      filters: { maturity: ['graduated'] },
    },
  },
  {
    name: 'Only incubating projects',
    filters: {
      pageNumber: 1,
      filters: { maturity: ['incubating'] },
    },
  },
  {
    name: 'Only sandbox projects',
    filters: {
      pageNumber: 1,
      filters: { maturity: ['sandbox'] },
    },
  },
  {
    name: 'Projects with A rating',
    filters: {
      pageNumber: 1,
      filters: { rating: ['a'] },
    },
  },
  {
    name: 'Projects with B rating',
    filters: {
      pageNumber: 1,
      filters: { rating: ['b'] },
    },
  },
  {
    name: 'Projects with C rating',
    filters: {
      pageNumber: 1,
      filters: { rating: ['c'] },
    },
  },
  {
    name: 'Projects with D rating',
    filters: {
      pageNumber: 1,
      filters: { rating: ['d'] },
    },
  },
  {
    name: 'Projects accepted by CNCF',
    filters: {
      pageNumber: 1,
      filters: { foundation: ['cncf'] },
    },
  },

  {
    name: 'Projects accepted by LF AI & Data',
    filters: {
      pageNumber: 1,
      filters: { category: ['lfaidata'] },
    },
  },
];

const QUERIES_NUMBER = 5;

const SampleQueries = (props: Props) => {
  const queries = QUERIES.length > QUERIES_NUMBER ? sampleSize(QUERIES, QUERIES_NUMBER) : QUERIES;

  return (
    <>
      {queries.map((query: SampleQuery, index: number) => (
        <Fragment key={`sampleQuery_${index}`}>
          <Link
            className={`badge rounded-0 border fw-normal mx-2 mt-3 text-decoration-none ${props.className}`}
            to={{
              pathname: '/search',
              search: prepareQueryString(query.filters),
            }}
            aria-label={`Filter by ${query.name}`}
          >
            {query.name}
          </Link>
          {!isUndefined(props.lineBreakIn) && index === props.lineBreakIn - 1 && (
            <div className="d-block w-100" data-testid="sampleQueryBreakLine" />
          )}
        </Fragment>
      ))}
    </>
  );
};

export default memo(SampleQueries);
