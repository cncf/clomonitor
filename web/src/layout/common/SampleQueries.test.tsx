import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { BrowserRouter as Router } from 'react-router-dom';

import SampleQueries from './SampleQueries';

const mockQueries = [
  {
    name: 'Only graduated projects',
    filters: {
      pageNumber: 1,
      filters: { maturity: ['graduated'] },
    },
  },
  {
    name: 'Only incubating projects',
    filters: {
      pageNumber: 1,
      filters: { maturity: ['incubating'] },
    },
  },
  {
    name: 'Only sandbox projects',
    filters: {
      pageNumber: 1,
      filters: { maturity: ['sandbox'] },
    },
  },
];

jest.mock('lodash', () => ({
  ...(jest.requireActual('lodash') as {}),
  sampleSize: () => {
    return mockQueries;
  },
}));

describe('SampleQueries', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const { asFragment } = render(
      <Router>
        <SampleQueries />
      </Router>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  describe('Render', () => {
    it('renders component', () => {
      render(
        <Router>
          <SampleQueries />
        </Router>
      );

      expect(screen.getAllByRole('link', { name: /Filter by/ })).toHaveLength(mockQueries.length);

      for (let i = 0; i < mockQueries.length; i++) {
        expect(screen.getByText(mockQueries[i].name)).toBeInTheDocument();
      }
    });

    it('renders proper classes', () => {
      render(
        <Router>
          <SampleQueries className="badge-light border-secondary text-secondary" />
        </Router>
      );

      const links = screen.getAllByRole('link', { name: /Filter by/ });
      expect(links[0]).toHaveClass('badge-light border-secondary text-secondary');
    });

    it('renders break line', () => {
      render(
        <Router>
          <SampleQueries lineBreakIn={2} />
        </Router>
      );

      expect(screen.getByTestId('sampleQueryBreakLine')).toBeInTheDocument();
    });

    it('opens first sample query', () => {
      render(
        <Router>
          <SampleQueries />
        </Router>
      );

      const links = screen.getAllByRole('link', { name: /Filter by/ });
      userEvent.click(links[0]);

      expect(window.location.pathname).toBe('/search');
      expect(window.location.search).toBe('?maturity=graduated&page=1');
    });
  });
});
