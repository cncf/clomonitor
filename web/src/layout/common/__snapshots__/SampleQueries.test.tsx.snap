// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SampleQueries creates snapshot 1`] = `
<DocumentFragment>
  <a
    aria-label="Filter by Only graduated projects"
    class="badge rounded-0 border fw-normal mx-2 mt-3 text-decoration-none undefined"
    href="/search?maturity=graduated&page=1"
  >
    Only graduated projects
  </a>
  <a
    aria-label="Filter by Only incubating projects"
    class="badge rounded-0 border fw-normal mx-2 mt-3 text-decoration-none undefined"
    href="/search?maturity=incubating&page=1"
  >
    Only incubating projects
  </a>
  <a
    aria-label="Filter by Only sandbox projects"
    class="badge rounded-0 border fw-normal mx-2 mt-3 text-decoration-none undefined"
    href="/search?maturity=sandbox&page=1"
  >
    Only sandbox projects
  </a>
</DocumentFragment>
`;
